HaskellEmacs.hs

```diff
@@ -135,17 +139,43 @@ allExports :: [Text] -> (Text, [Text])
 allExports = g . filter (not . null) . map exportsGet
   where g = T.unlines . map head &&& concatMap tail
 
--- | Retrieve list of exported functions in a haskell module.
-exportsGet :: Text -> [Text]
-exportsGet t
-  | length list < 2 = []
-  | otherwise       = (\(x:xs) -> imports x : map ((x <> ".") <>) xs) list
-  where list = filter (not . T.null) . takeWhile (/= "where")
-               . drop 1 . dropWhile (/= "module") $ T.split (`elem` ("\n ,()\t" :: String))
-               . T.unlines . map (fst . T.breakOn "--") $ T.lines t
-        imports = (<>) "import qualified "
-
 -- | List of haskell functions which get querried for their arity.
 arityFormat :: [Text] -> Text
 arityFormat = T.intercalate ","
               . map (\x -> "(\"" <> x <> "\",arity " <> x <> ")")
+
+-- | Retrieve list of exported functions in a haskell module.
+exportsGet :: Text -> [Text]
+exportsGet moduleContent =
+  case parseModule (unpack moduleContent) of
+    ParseOk (Module _ _ _ _ Nothing _ decls)    -> exportsFromModuleDecls decls
+    ParseOk (Module _ _ _ _ (Just exspecs) _ _) -> exportsFromHeader exspecs
+    ParseFailed loc msg                         -> error msg
+    
+exportsFromModuleDecls :: [Decl] -> [Text]
+exportsFromModuleDecls = catMaybes . fmap functionDeclarationNames
+
+functionDeclarationNames :: Decl -> Maybe Text
+functionDeclarationNames (FunBind [])                         = Nothing
+functionDeclarationNames (FunBind ((Match _ nm _ _ _ _) : _)) = Just (fromName nm)
+functionDeclarationNames (PatBind _ (PVar nm) _ _)            = Just (fromName nm)
+functionDeclarationNames _                                    = Nothing 
+
+-- Extract the unqalified function names from an ExportSpec  
+exportsFromHeader :: [ExportSpec] -> [Text]
+exportsFromHeader =
+  catMaybes . fmap (fmap fromName  . exportFunction)
+
+fromName :: Name -> Text
+fromName (S.Symbol str) = pack str
+fromName (S.Ident str)  = pack str
+
+exportFunction :: ExportSpec -> Maybe Name
+exportFunction (EVar _ qname)      = unQalifiedName qname
+exportFunction (EModuleContents _) = Nothing
+exportFunction _                   = Nothing
+
+unQalifiedName :: QName -> Maybe Name
+unQalifiedName (Qual _ name) = Just name
+unQalifiedName (UnQual name) = Just name
+unQalifiedName _             = Nothing
```

Comment:
Will only use SYB stlye generic programming if boiler plate starts to get completely out of control.